File: design_patterns/C++/Creational/factoryMethod.cxx
#include <iostream>

class Interviewer {
  public:
    virtual void askQuestions() = 0;
};

class Developer: public Interviewer {
  public:
    void askQuestions() {
        std::cout << "Asking about design patterns!" << std::endl;
    }
};

class CommunityExecutive: public Interviewer {
  public:
    void askQuestions() {
        std::cout << "Asking about community building!" << std::endl;
    }
};

class HiringManager {
  public:
    virtual Interviewer* makeInterviewer() = 0;  // factory method

    void takeInterview() {
        Interviewer* interviewer = this->makeInterviewer();
        interviewer->askQuestions();
    }
};

class DevelopmentManager: public HiringManager {
  public:
    Interviewer* makeInterviewer() {
        return new Developer();
    }
};

class MarketingManager: public HiringManager {
  public:
    Interviewer* makeInterviewer() {
        return new CommunityExecutive();
    }
};

int main() {
    auto devManager = new DevelopmentManager();
    auto marketingManager = new MarketingManager();
    devManager->takeInterview();
    marketingManager->takeInterview();
    // ================================
    // Asking about design patterns!
    // Asking about community building!
    // ================================
    return 0;
}

File: design_patterns/Python/Creational/factory_method.py
class Interviewer:

    @staticmethod
    def ask_questions():
        raise NotImplementedError


class Developer(Interviewer):

    def ask_questions(self):
        print('Asking about design patterns!')


class CommunityExecutive(Interviewer):

    def ask_questions(self):
        print('Asking about community building')


class HiringManager:

    def make_interviewer(self):
        raise NotImplementedError

    def take_interview(self):
        interviewer = self.make_interviewer()
        interviewer.ask_questions()


class DevelopmentManager(HiringManager):

    def make_interviewer(self):
        return Developer()


class MarketingManager(HiringManager):

    def make_interviewer(self):
        return CommunityExecutive()


if __name__ == '__main__':

    dev_manager = DevelopmentManager()
    dev_manager.take_interview()
    # Output: Asking about design patterns!

    mkt_manager = MarketingManager()
    mkt_manager.take_interview()
    # Output: Asking about community building
